### PR TITLE
Issue #33: Write `add_event_vars` preprocessing functionality

### PR DESCRIPTION
**Pull request by athowes**
## Description

This PR will close issue #33. The goal is to add a modular function which adds interval censoring columns.

* Adds a function `add_event_vars`
* This function:
  1. Gets the columns of `linelist` renamed into our preferred names
  2. Allows users to pass `pwindow` and `swindow` as either charater point to the name of the column, or numeric in which case the column will be created
  3. Creates the `lwr` `upr` and `window` columns from other columns if names pointing to them not provided 
  4. Checks that the columns are numeric, above zero, and add up to equal each other when relevant
  5. Rearranges the columns
* Adds a helper function `.rename_column` which renames the column if provided two strings and the dataframe
* This function does not yet slot into anywhere within the code. I have left `observe_process()` alone for now. Eventually this function plus others will slot in to replace it

About reviewing this PR:

* I think I might have made the function a bit overly complicated so any help reducing the complexity appreciated
* There are also some cases given the amount of complexity where I am not 100% sure it does what we want

Remains to do:

* Document `add_event_vars` better (i.e. the parameters have no description at the moment)
* Document `.rename_column` and maybe add tests for it

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [ ] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

**Comment by seabbs** [comment]
> I think I might have made the function a bit overly complicated so any help reducing the complexity appreciated

I haven't read the code yes but agree from reading the overview. My current uninformed view is that this is really a few functions combined. 

I think some clarity here could be had in thinking about what exactly is happening here and giving that names

File: R/preprocess.R [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    linelist$pwindow <- pwindow
+  }
+
+  if (is.numeric(swindow)) {
+    cli::cli_warn("Overwriting using numeric value(s) of swindow provided!")
+    linelist$swindow <- swindow
+  }
+
+  if (is.null(stime_upr)) {
+    linelist <- mutate(linelist, stime_upr = stime_lwr + swindow)
```
Comment (seabbs) [inline R/preprocess.R:35]: Performance doesn't matter hugely but noting as written this is a very inefficient function as every rename, mutate, etc all here leads to a data copy.

File: R/preprocess.R [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#' @param ptime_upr ...
+#' @param pwindow ...
+#' @param stime_lwr ...
+#' @param stime_upr ...
+#' @param swindow ...
+#' @family preprocess
+#' @autoglobal
+#' @export
+add_event_vars <- function(
+  linelist, ptime_lwr = NULL, ptime_upr = NULL, pwindow = NULL,
```
Comment (seabbs) [inline R/preprocess.R:14]: I think adding event vars is really going from having p time in some format and s time in some format (lower bound assumed) and adding pwindow and window to get the event variables. 

I think the step of making sure we have _lwr and upr for each is another function (and would perhaps naturally live in a general as_epidist_linelist class which can then optionally be passed to as_latent_model. 

If not that s3 design then something else.

**Review by seabbs — COMMENTED** [review]
I added some thoughts after looking at the code. As I said I think this function is trying to do too much. One option I have suggested is an initial s3 class as_epidist_linelist we can then convert into model formats which might be nice. 

A path here is we accept this PR nearly as is and write a new issue to clean up. Another option is we simplify the functionality here and address the rest later.

**Comment by athowes** [comment]
Yep agree want more clarity here.

I was also looking at `as_latent_individual` and I think even though we can't replace `observe_process` we might be able to start getting rid of bits of `as_latent_individual` within these PRs.

**Review by seabbs — APPROVED** [review]
Accepting under the assumption is will shortly be refactored. Noting that the partner issue is missing detail of the pr conversation here

**Pull request merged**